File: 98-validate-binary-search-tree/validate-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 #define lint long int
class Solution {
public:
    bool helper(TreeNode* root, lint minm, lint maxm){
        if(!root) return true;
        return (lint)root->val>minm && (lint)root->val<maxm && helper(root->left, minm, root->val) && helper(root->right, root->val, maxm); 
    }
    bool isValidBST(TreeNode* root) {
        lint maxm=INT_MAX, minm=INT_MIN;
        maxm++, minm--;
        return helper(root, minm, maxm);
    }
};

File: 98-validate-binary-search-tree/validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private boolean isValidBSTUtil(TreeNode root, long minValue, long maxValue) {
        if(root==null) return true;
        long val= root.val;
        return val>minValue && val<maxValue
                && isValidBSTUtil(root.left, minValue, val) 
                    && isValidBSTUtil(root.right, val, maxValue);
    }
    public boolean isValidBST(TreeNode root) {
        return isValidBSTUtil(root, Long.MIN_VALUE, Long.MAX_VALUE);
    }
}